WalletWasabi.Tests/UnitTests/Blockchain/TransactionBuilding/BranchAndBoundTests.cs

```diff
@@ -74,6 +74,44 @@ public void MoreSelection_ExactMatchIsExpensive()
 		Assert.Equal(new long[] { 35 }, actualSelection);
 	}
 
+	/// <summary>
+	/// The goal of the test is to verify that we prune branches at points when it is clear that
+	/// the sum of remaining coins is less than what we need to hit our target.
+	/// </summary>
+	/// <remarks>
+	/// This is especially important for cases where user has many coins.
+	/// <para>If the optimization is not in place, you should observe that this test does not really finish fast.</para>
+	/// </remarks>
+	[Fact]
+	public void MoreSelection_RemainingAmountOptimization()
+	{
+		List<long> inputValues = new();
+		inputValues.Add(1_000_000);
+
+		for (int i = 0; i < 1000; i++)
+		{
+			inputValues.Add(1);
+		}
+
+		// All inputs cost the same.
+		long[] inputCosts = inputValues.Select(x => 1L).ToArray();
+		
+		long target = 999_999; // Target that we cannot get as a sum of input values.
+
+		BranchAndBound algorithm = new();
+		MoreSelectionStrategy strategy = new(target, inputValues.ToArray(), inputCosts);
+		bool wasSuccessful = algorithm.TryGetMatch(strategy, out List<long>? selectedCoins);
```

Comment:
Maybe we could have a `CancellationTokenSource` with 15 sec timeout (or less) here to be sure the algo doesn't time out, because we have the "remaining amount optimization" in place.

```suggestion
		using CancellationTokenSource cts = new(TimeSpan.FromSeconds(15));

		BranchAndBound algorithm = new();
		MoreSelectionStrategy strategy = new(target, inputValues.ToArray(), inputCosts);
		bool wasSuccessful = algorithm.TryGetMatch(strategy, out List<long>? selectedCoins, cts.Token);
```

You will need to grab the latest `master` for this suggestion to make sense. (#7332 got merged)

Reply:
I think that having a `CancellationTokenSource` here is a good idea. It's easier to find out that a test timed out rather than that it never finished. So 👍. 

I would say it's better to use a larger rather than lower timeout, so I used `30 seconds` instead of `15` just because we don't have any guarrantees on CI regarding its performance. Even though in this case it does not matter too much because we should never hit either 15 or 30 seconds timeout because we believe the optimization is correctly implemented. If it is not, we will look into it and then it's again not important if it times out in 15 or 30 seconds.